### PR TITLE
refactor: switch to Typescript (incl. tslint, types, tests)

### PR DESCRIPTION
**Pull request by kyegupov**
- [x] Ready for review
- [x] Follows CONTRIBUTING rules
- [x] Reviewed by Snyk internal team

#### What does this PR do?

Converts the project to Typescript.
Includes proper linting and type definitions.

#### Where should the reviewer start?

Look at index.ts for high-level definitions and package.json changes for infrastructure.

#### How should this be manually tested?

npm link from Snyk CLI

#### Any background context you want to provide?

This is a redoing of botched https://github.com/snyk/snyk-gradle-plugin/pull/34

#### What are the relevant tickets?

Motivated by https://snyksec.atlassian.net/browse/BST-508

### CODE REVIEW

File: lib/init.gradle [hunk 1]
```diff
@@ -18,22 +18,26 @@ import groovy.json.JsonOutput
...
-//     dependencies?: DepDict;
+//   name: string;
+//   version: string;
+//   dependencies?: DepDict;
 // }
 
 rootProject { rootProj ->
     task snykResolvedDepsJson { 
-        def result = ['$rootProject': rootProj.name]
+        def projectsDict = [:]
```
Comment (miiila) [inline lib/init.gradle]: If I read it correctly, this is also changing behaviour of `snykResolvedDepsJson`, right? Should this be commited in refactor?

File: README.md [hunk 2]
```diff
@@ -21,3 +21,10 @@ Additional command line arguments to Gradle can be provided after `--`, for exam
 ## Under the hood
 
 See `lib/init.gradle` for the Groovy script injected in Gradle builds to gather and resolve the dependencies.
+
+## Developer notes
+
+Minimum supported version of Node.js is 4.
+
+It does not support ts-node (required to run tests from `.ts` files), so we have to run `.js` tests from `./dist`
```
Comment (miiila) [inline README.md]: This doesn't sound right as we use tap with ts-node on node 4; https://github.com/snyk/snyk/blob/master/package.json#L30 and https://travis-ci.org/snyk/snyk/jobs/501734721#L499

File: tsconfig.json [hunk 3]
```diff
@@ -0,0 +1,21 @@
...
+      "pretty": true,
+      "target": "es5",
+      "lib": [
+        "es5",
+        "es2015.promise",
+      ],
+      "module": "commonjs",
+      "allowJs": false,
+      "sourceMap": true,
+      "strict": true,
```
Comment (miiila) [inline tsconfig.json]: I'd add `"declaration": true,` for easier usage in typescript codebase.

File: package.json [hunk 4]
```diff
@@ -6,26 +6,31 @@
...
   "scripts": {
-    "test": "npm run lint && npm run test-functional && npm run test-system",
-    "lint": "eslint -c .eslintrc lib test",
-    "test-functional": "tap -R spec ./test/functional/*.test.js",
-    "test-system": "tap -R spec --timeout=180 ./test/system/*.test.js",
+    "build": "tsc",
+    "test": "npm run build && npm run lint && npm run test-functional && npm run test-system",
+    "lint": "tslint --project tsconfig.json --format stylish",
+    "test-functional": "tap -R spec ./dist/test/functional/*.test.js",
+    "test-system": "tap -R spec --timeout=180 ./dist/test/system/*.test.js",
```
[the remark this answers is not in the document]
Comment (kyegupov) [reply package.json]: we also need init.gradle

File: package.json [hunk 5]
```diff
@@ -6,26 +6,31 @@
...
+  "main": "dist/index.js",
   "directories": {
     "test": "test"
   },
   "scripts": {
-    "test": "npm run lint && npm run test-functional && npm run test-system",
-    "lint": "eslint -c .eslintrc lib test",
-    "test-functional": "tap -R spec ./test/functional/*.test.js",
-    "test-system": "tap -R spec --timeout=180 ./test/system/*.test.js",
+    "build": "tsc",
```
[the remark this answers is not in the document]
Comment (kyegupov) [reply package.json]: Done.

File: package.json [hunk 6]
```diff
@@ -6,26 +6,31 @@
     "type": "git",
     "url": "https://github.com/snyk/snyk-gradle-plugin"
   },
-  "main": "lib/index.js",
+  "main": "dist/index.js",
```
[the remark this answers is not in the document]
Comment (kyegupov) [reply package.json]: oh yes.

File: package.json [hunk 7]
```diff
@@ -6,26 +6,31 @@
...
   "author": "snyk.io",
   "license": "Apache-2.0",
   "devDependencies": {
-    "eslint": "^4.11.0",
+    "@types/node": "^4.9.1",
+    "@types/sinon": "^7.0.10",
     "semantic-release": "^15",
     "sinon": "^2.4.1",
-    "tap": "github:snyk/node-tap#alternative-runtimes"
+    "tap": "github:snyk/node-tap#alternative-runtimes",
```
[the remark this answers is not in the document]
Comment (kyegupov) [reply package.json]: But does it work with node 4? I can test, but I'd rather do it in a separate change. The current way works.

File: lib/init.gradle (hunk 1 above)
Comment (kyegupov) [reply lib/init.gradle]: It is technically a refactor still :)
Without it, type definitions for the output are harder to define.

File: README.md (hunk 2 above)
Comment (kyegupov) [reply README.md]: Doesn't work for me: .ts tests are not getting picked up.
I'd rather not sink any time into it 
Reworded to be more true.

File: tsconfig.json (hunk 3 above)
Comment (kyegupov) [reply tsconfig.json]: Not sure if it adds a lot of value (the declarations already are visible in .ts files)

Comment (miiila) [reply tsconfig.json]: But when you distribute compiled part only, you don't provide .ts files. So if you require that module in TS code, you'd miss the type definitions. It's more about distributing JS, but being nice to TS developers in a same time.

**Pull request merged**